lib/Core/PTree.h

```diff
@@ -28,14 +39,26 @@ namespace klee {
   };
 
   class PTree {
+    // Number of registered ID
+    int registeredIds = 0;
+
   public:
-    std::unique_ptr<PTreeNode> root;
+    PTreeNodePtr root;
     explicit PTree(ExecutionState *initialState);
     ~PTree() = default;
 
-    static void attach(PTreeNode *node, ExecutionState *leftState, ExecutionState *rightState);
+    void attach(PTreeNode *node, ExecutionState *leftState,
+                ExecutionState *rightState);
     static void remove(PTreeNode *node);
     void dump(llvm::raw_ostream &os);
+    uint8_t getNextId() {
+      auto id = 1 << registeredIds++;
```

Comment:
Why `auto`? Doesn't it default to `int` here?

Reply:
Because I didn't want to think about what is the appropriate type. Isn't that the point of `auto` in this case? I think I was hopping it would infer the type from the return type of the function.

Reply:
No, this is not Haskell. C++ just sees the 1 and chooses `int`. You could use `decltype` here but @ccadar might hate it. :stuck_out_tongue: Just use `std::uint8_t`...

Reply:
Well clearly the compiler needs to get better!

Reply:
Or the C++ standard... However, you can declare the return type as auto: `auto getNextId() {...}`. This is allowed since C++14. :grin: